src/Storages/S3Queue/S3QueueSource.cpp

```
@@ -218,61 +118,86 @@ String StorageS3QueueSource::getName() const
+        auto file_status = key_with_info->file_status;
+
+        auto * prev_scope = CurrentThread::get().attachProfileCountersScope(&file_status->profile_counters);
+        SCOPE_EXIT({ CurrentThread::get().attachProfileCountersScope(prev_scope); });
+        /// FIXME:  if files are compressed, profile counters update does not work fully (s3 related counters are not saved). Why?
+
         try
         {
+            auto timer = DB::CurrentThread::getProfileEvents().timer(ProfileEvents::S3QueuePullMicroseconds);
+
+            Chunk chunk;
             if (reader->pull(chunk))
             {
-                UInt64 num_rows = chunk.getNumRows();
-                auto file_path = reader.getPath();
-
-                for (const auto & virtual_column : requested_virtual_columns)
-                {
-                    if (virtual_column.name == "_path")
-                    {
-                        chunk.addColumn(virtual_column.type->createColumnConst(num_rows, file_path)->convertToFullColumnIfConst());
-                    }
-                    else if (virtual_column.name == "_file")
-                    {
-                        size_t last_slash_pos = file_path.find_last_of('/');
-                        auto column = virtual_column.type->createColumnConst(num_rows, file_path.substr(last_slash_pos + 1));
-                        chunk.addColumn(column->convertToFullColumnIfConst());
-                    }
-                }
-                success_in_pulling = true;
+                LOG_TEST(log, "Read {} rows from file: {}", chunk.getNumRows(), reader.getPath());
+
+                file_status->processed_rows += chunk.getNumRows();
+                processed_rows_from_file += chunk.getNumRows();
+
+                VirtualColumnUtils::addRequestedPathAndFileVirtualsToChunk(chunk, requested_virtual_columns, reader.getPath());
+                return chunk;
             }
         }
         catch (const Exception & e)
```

I think we should catch all exceptions, not only `DB::Exception` here